src/web/debugger/session.ts

```diff
@@ -0,0 +1,296 @@
+        this.driver.resume(SimulatorDebuggerCommand.StepOut);
+        this.sendResponse(response);
+    }
+
+    protected pauseRequest(response: DebugProtocol.PauseResponse, args: DebugProtocol.PauseArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.Pause);
+        this.sendResponse(response);
+    }
+
+    protected threadsRequest(response: DebugProtocol.ThreadsResponse): void {
+        response.body = { threads: [{ id: SimDebugSession.THREAD_ID, name: "main" }] }
+        this.sendResponse(response);
+    }
+
+    protected stackTraceRequest(response: DebugProtocol.StackTraceResponse, args: DebugProtocol.StackTraceArguments): void {
+        if (this.lastBreak && this.state) {
+            const frames = this.state.getFrames();
+            response.body = { stackFrames: frames };
+        }
+        this.sendResponse(response);
+    }
+
+    protected scopesRequest(response: DebugProtocol.ScopesResponse, args: DebugProtocol.ScopesArguments): void {
+        if (this.state) {
+            response.body = { scopes: this.state.getScopes(args.frameId) }
+        }
+
+        this.sendResponse(response);
+    }
+
+    protected async variablesRequest(response: DebugProtocol.VariablesResponse, args: DebugProtocol.VariablesArguments) {
+        if (this.state) {
+            const variables = await this.state.getVariables(args.variablesReference);
+            response.body = { variables };
+        }
+
+        this.sendResponse(response);
+    }
+
+    private onDebuggerBreakpoint(breakMsg: pxsim.DebuggerBreakpointMessage) {
+        this.lastBreak = breakMsg;
+        this.state = new StoppedState(this.lastBreak, this.breakpoints!, this.driver);
+
+        if (breakMsg.exceptionMessage) {
+            const message = breakMsg.exceptionMessage.replace(/___\d+/g, '');
+            this.sendEvent(new StoppedEvent("exception", SimDebugSession.THREAD_ID, message));
+        }
+        else {
+            this.sendEvent(new StoppedEvent("breakpoint", SimDebugSession.THREAD_ID));
+        }
+    }
+
+    private onDebuggerWarning(warnMsg: pxsim.DebuggerWarningMessage) {
+    }
+
+    private onDebuggerResume() {
+        this.sendEvent(new ContinuedEvent(SimDebugSession.THREAD_ID, true));
+    }
+
+    private onStateChanged(state: SimulatorState) {
+        switch (state) {
+            case SimulatorState.Paused:
+                // Sending a stopped event here would be redundant
+                break;
+            case SimulatorState.Running:
+                this.sendEvent(new ContinuedEvent(SimDebugSession.THREAD_ID, true))
+                break;
+            case SimulatorState.Stopped:
+                this.sendEvent(new TerminatedEvent())
+                break;
+            //case SimulatorState.Unloaded:
+            //case SimulatorState.Pending:
+            default:
+        }
+    }
+
+    private fixBreakpoints() {
```

Comment:
nit: something like `convertBreakpoints` might make the intent of the function clearer